trieste/models/keras/architectures.py

```diff
@@ -233,3 +235,148 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
         output_tensor = self._gen_output_layer(hidden_tensor)
 
         return input_tensor, output_tensor
+
+
+class DropoutNetwork(KerasEnsembleNetwork):
+    """
+    This class builds a standard dropout neural network using Keras. The network
+    architecture is a multilayer fully-connected feed-forward network, with Dropout layers
+    preceding each fully connected dense layer. The network is meant to be passed to 
+    :class:`MCDropout` which will define the predict method to make this a probabilistic model.
+    """
+
+    def __init__(
+        self,
+        input_tensor_spec: tf.TensorSpec,
+        output_tensor_spec: tf.TensorSpec,
+        hidden_layer_args: Sequence[dict[str, Any]] = (
+            {"units": 50, "activation": "relu"},
+            {"units": 50, "activation": "relu"},
+            {"units": 50, "activation": "relu"},
+        ),
+        rate: Sequence[float] | float = 0.5
+    ):
+        """
+        :param input_tensor_spec: Tensor specification for the input to the network.
+        :param output_tensor_spec: Tensor specification for the output of the network.
+        :param hidden_layer_args: Specification for building dense hidden layers. Each element in
+            the sequence should be a dictionary containing arguments (keys) and their values for a
+            :class:`~tf.keras.layers.Dense` hidden layer. Please check Keras Dense layer API for
+            available arguments. Objects in the sequence will sequentially be used to add
+            :class:`~tf.keras.layers.Dense` layers. Length of this sequence determines the number of
+            hidden layers in the network. Default value is two hidden layers, 50 nodes each, with
+            ReLu activation functions. Empty sequence needs to be passed to have no hidden layers.
+        :param rate: Probability of dropout assigned to each layer of a fully connected network.
+            If scalar, the same probability will be assigned to each layer. Otherwise accepts a list of 
+            probabilities corresponding to each layer. 
+        :raise ValueError: If objects in ``hidden_layer_args`` are not dictionaries.
+        """
+        super().__init__(input_tensor_spec, output_tensor_spec)
+
+        self._hidden_layer_args = hidden_layer_args
+        self.rate = rate
+        self._model = self._build_model()
+
+    @property
+    def model(self) -> tf.keras.Model:
+        return self._model
+
+    @property
+    def rate(self) -> Sequence[float]:
+        return self._rate
+    
+    @rate.setter
```

Comment:
is rate meant to be changed after the model is instantiated?

Reply:
Probably not. I was more aiming for giving the flexibility of setting the rate with either a single float or a list of floats, if for some reason it was desirable to pass different probabilities for example. 

The setting of rates has been used as convenience in some of the tests which could be written to avoid it.

Reply:
but that would still be passed when the object/network is created, no? it seems to me that regardless of whether you give a single float or a list you wouldn't change it afterwards so this setter is not needed

if you don't see a clear need for this additional flexibility its better to simplify, if really needed this can be amended later on